`has_triton`: Use the device interface for detecting Triton availability

galexite opened the pull request:
This PR replaces the `has_triton()` global method which was previously used for this task.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @kwen2501 @c-p-i-o @yf225 @ColinPeppler @desertfire @rec

galexite commented:
@pytorchbot label 'topic: not user facing'

galexite commented:
@jansel sorry about that, looks like I messed up the typing, should be ready for another CI run now!

jansel commented:
@pytorchbot merge

galexite commented:
@jansel @tugsbayasgalan I've fixed the tests and added the HOPs from Inductor that appear when I call `init_backends_registration` (which then eventually imports `torch._inductor.ir`) in `inductor_utils.py` to the list of HOPs without op info here: https://github.com/pytorch/pytorch/pull/139171/files#diff-2a1edf2b2655350ac32e1d24b4cfaf5b65d90056c4fcb2f77600245e428d5131R78-R84. I hope this is okay.

galexite commented:
Hey @jansel, could I trouble you for a re-review please? Thanks!

jansel reviewed (changes requested):
Failing tests?

galexite commented:
> Failing tests?

Yep, I managed to let a `has_triton()` slip in through the merge that I pushed after that comment... that has now been fixed.

jansel reviewed (changes requested):
Lints are failing now

jansel commented:
@pytorchbot merge

jansel commented:
@galexite that test failure looks real

jansel commented:
Test failure is `RuntimeError: Found Tesla M60 which is too old to be supported by the triton GPU compiler, which is used as the backend. Triton only supports devices of CUDA Capability >= 7.0, but your device is of CUDA capability 5.2`

jansel commented on torch/_inductor/codegen/cuda_combined_scheduling.py:
```diff
@@ -30,8 +34,15 @@ def __init__(self, scheduler: Scheduler) -> None:
+    def check_if_available(
+        cls, device: "Union[str, torch.device, None]" = None
+    ) -> None:
+        TritonScheduling.check_if_available(device)
```
```suggestion
        return TritonScheduling.check_if_available(device)
```

jansel replied:
We should rename the function to something like `assert` or `raise_if`

jansel commented on test/inductor/test_aot_inductor_arrayref.py:
```diff
@@ -72,7 +72,6 @@ def fail_minimal_arrayref_interface(is_skip=False):
     "test_cond_with_reinterpret_view_inputs_outputs": fail_minimal_arrayref_interface(),
     "test_cond_share_predicte": fail_stack_allocation(is_skip=True),
     "test_while_loop_with_parameters": fail_minimal_arrayref_interface(),
-    "test_while_loop_with_pytree_inputs": fail_stack_allocation(),
```
Why does this PR fix this issue?

galexite replied:
I suspect this was because of calling `init_backend_registration()` in `inductor_utils.py` (from my understanding of the original PR that added this xfail). I'll wait and see what CI says, but now this is lazily done the failure should return.

jansel replied:
Can you split this fix into a different PR?

galexite replied:
I don't think this failure has returned, and I certainly don't think I actually 'fixed' this, so I'll convert this xfail in to a skip.

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -3679,6 +3681,18 @@ def get_backend_features(cls, device: torch.device):
+        if device is None:
+            device = torch.get_default_device()
+
+        get_interface_for_device(device).check_if_triton_available(device)
```
```suggestion
        return get_interface_for_device(device).check_if_triton_available(device)
```

jansel commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -1502,9 +1500,6 @@ def decorator(func):
         def wrapper(*args, **kwargs):
             original_fully_shard = torch.distributed.fsdp.fully_shard
             for mode in FullyShardMode:
-                if mode != FullyShardMode.EAGER and not has_triton():
-                    warnings.warn("Inductor on GPU needs Triton and recent GPU arch")
-                    continue
```
Why is this being removed?

jansel commented on torch/testing/_internal/hop_db.py:
Revert unrelated changes in this file.  If they are intentional put them in a different PR.

galexite replied:
I think these are required because of the imports that I've added in to `inductor_utils.py` bring these HOPs back in to view, but I'll give removing it a go.

jansel commented on torch/testing/_internal/inductor_utils.py:
```diff
@@ -22,9 +25,11 @@
 
 log: logging.Logger = logging.getLogger(__name__)
 
+
```
Do a pass to revert unrelated changes.

jansel commented on torch/testing/_internal/inductor_utils.py:
```diff
@@ -37,22 +42,60 @@ def test_cpu():
-else:
-    TRITON_HAS_CPU = False
+# Ensure the scheduling backends are registered first.
+init_backend_registration()
```
This should be done lazily not at import time since it will prevent people from overriding.

EikanWang commented on torch/_inductor/fx_passes/pad_mm.py:
```diff
@@ -440,7 +442,11 @@ def realize_symbols(ds):
             return True
 
-        if not has_triton():
+        scheduling_factory = get_scheduling_for_device(mat1.device.type)
```
In addition to the existence of the Triton package, `has_triton()` requires at least one GPU device to be available. However, `get_scheduling_for_device(mat1.device.type)` only requires the existence of the Triton package. It means that `CPU` will run into this pass as well.

Meanwhile, the Triton package will be installed automatically if users install CUDA, Rocm,, or XPU(Intel GPU) torch packages even if a real GPU device does not exist, not mention to the CUDA package is the default wheel package of torch pip channel.

EikanWang replied:
Make sense. Thanks

EikanWang commented on torch/_dynamo/device_interface.py:
```diff
@@ -220,6 +242,29 @@ def get_compute_capability(device: _device_t = None):
+        )
+
+    @staticmethod
+    def raise_if_triton_unavailable(device: _device_t = None) -> None:
```
It is kind of wired to throw an exception for this check. Meanwhile, this API has covered `is_triton_capable`. Where does the `is_triton_capable` need to be invoked?

galexite replied:
This was inspired by how `verify_dynamo` worked, with the original functions `check_if_inductor_supported` and `check_if_dynamo_supported` (which I've now renamed in this PR). An alternative API would be to return a `str | None`, but that means `None` would be the happy (as in, Triton is supported) case, which seems a bit odd.

Implementations of `raise_if_triton_unavailable` should call `is_triton_capable` – that method is only to check whether a device would actually be supported by Triton (as in CUDA below, CUDA on Triton requires a certain device architecture version), not whether we can actually use Triton on the device because of the current operating environment.

galexite commented:
Hey @jansel, sorry to bother you, this should be ready for review again if you have the chance. The XPU failures look unrelated, they seem to be showing up on other ciflow/xpu PRs as well.

jansel reviewed (changes requested):
We should either fix or skip those xpu failures.  main should be kept green, so if it is broken on main lets add a (different PR) to fix the errors.

galexite commented:
I've removed all the changes to tests, this PR was getting unwieldy, I'll submit more PRs to gradually move the tests over later.

galexite commented:
This should be ready for review again!

galexite commented:
@pytorchbot merge

galexite commented:
Had to add an exception to `_content_store.py` for MPS because `is_compile_supported()` didn't return `True` for MPS until this PR, and whilst there is handling of MPS below, MPS doesn't support some of the ops required for hashing.

galexite commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Sorry but this is breaking internally. @jansel can you please help get these changes working? See D70946254 for more details. To validate the fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"

Relevant error message appears to be: 
`RuntimeError: Could not save cache to remote as [path] does not exist. max_autotune is set to False.`

galexite commented:
@jansel wrt to @ZainRizvi's comment above for the revert, I'm not really sure where to look for this, I think I might have inadvertently enabled/disabled Triton somewhere I shouldn't have?

shink commented:
Many thanks for your work on this feature! When can we expect this change to be relanded?

galexite commented:
Hi @shink! I'd like to land this, but I'm afraid I don't know the cause of the fb-internal failure which lead to the revert.

jansel commented:
If you want to rebase we can try importing it again.  You could also try breaking this change up into smaller pieces to help isolate the issue.

galexite commented:
I have split up the device interface component, which should merge fine because it's not used in any paths as-is, to #152529. Unfortunately, I don't have access to use ghstack, so I'll have to put up PRs separately and manually but I should be able to get it across the line.

galexite committed:
Use the device interface to back `has_triton`

galexite commented:
Rebased atop https://github.com/pytorch/pytorch/pull/152529.

I have removed the changes to the Inductor scheduler checks, I think that is where the problem may lie, instead this only includes the `has_triton` component. I'll submit the Inductor scheduler checks as a separate PR.

galexite committed:
Undo changes to _content_store.py, they are not necessary with these changes

galexite committed:
Undo changes to verify_dynamo.py

galexite committed:
Remove comment change from logging.py

galexite commented:
@pytorchbot merge

masnesral commented:
Looks like this causes a pretty big perf drop on some huggingface models. For example, on an H100 from 2.6x -> 1.3x for the following:
`python benchmarks/dynamo/huggingface.py --performance --inference --bfloat16 --backend inductor --device cuda --cold-start-latency --only BlenderbotSmallForCausalLM`

galexite commented:
Hi @masnesral, if this PR is causing those performance drops, it might be because Triton kernels aren't being used, because `has_triton` returns `False`? What happens if you do the following in a Python REPL?

```python
from torch.utils._triton import has_triton
print(f"{has_triton()=}")
print(f"{has_triton("cuda")=}")

from torch._dynamo.device_interface import get_interface_for_device
get_interface_for_device("cuda").raise_if_triton_unavailable()  # shouldn't throw if all is okay
```

Unfortunately, I don't have access to an H100, but this should work the same. These give `True` for me on a `g5.8xlarge` AWS instance and the last doesn't throw.

Also, to help with debugging this, could you also tell me what the result of these for you?
```
import triton, torch
print(f'{"nvidia" in triton.backends.backends=}')
print(f'{torch.cuda.get_device_properties("cuda")=}')
```

galexite commented:
The other thing is, I did change the `if` in Inductor which enables the `pad_mm` pass, to check if the tensor is on a device actively using the `TritonScheduler`, rather than using `has_triton`. Maybe that causes an early exit of that Inductor pass?

masnesral commented:
The perf drop is also pretty noticeable on A100. See for example:
https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Fri,%2002%20May%202025%2020:46:14%20GMT&stopTime=Fri,%2009%20May%202025%2020:46:14%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(h100)&lBranch=main&lCommit=aca2c99a6528349bb67481310c56a9b08f39934b&rBranch=main&rCommit=5c878d4b04be45c86ab280d0d5e33ba072c9dcb3

masnesral commented:
Sorry, gonna try to revert this while we investigate further.

masnesral commented:
@pytorchbot revert -m="Performance regression for huggingface" -c=nosignal

masnesral commented:
@galexite my having tested on H100 doesn't seem to be a related (see links to a100 above), but here's the output you requested anyway. I think it's what you were expecting :/

```
>>> import triton, torch
>>> print(f'{"nvidia" in triton.backends.backends=}')
"nvidia" in triton.backends.backends=True
>>> print(f'{torch.cuda.get_device_properties("cuda")=}')
torch.cuda.get_device_properties("cuda")=_CudaDeviceProperties(name='NVIDIA H100', major=9, minor=0, total_memory=97285MB, multi_processor_count=132, uuid=70fa10a6-2939-4471-959c-6da3b40decb6, pci_bus_id=6, pci_device_id=0, pci_domain_id=0, L2_cache_size=60MB)

>>> from torch.utils._triton import has_triton
>>> print(f"{has_triton()=}")
has_triton()=True
>>> print(f"{has_triton('cuda')=}")
has_triton('cuda')=True
>>> from torch._dynamo.device_interface import get_interface_for_device
>>> get_interface_for_device("cuda").raise_if_triton_unavailable()
```

galexite commented:
Hmm, okay. I'll have a look!

galexite committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into georgew/generalise-has-triton

galexite committed:
Move has_triton down to where it was originally

galexite committed:
Revert some changes that might cause performance regression

galexite committed:
Lint

galexite committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into georgew/generalise-has-triton

galexite committed:
Merge branch 'georgew/generalise-has-triton'
